src/OpenTelemetry.Adapter.Dependencies/Implementation/GrpcClientDiagnosticListener.cs

```diff
@@ -0,0 +1,115 @@
+    internal class GrpcClientDiagnosticListener : ListenerHandler
+    {
+        private readonly PropertyFetcher startRequestFetcher = new PropertyFetcher("Request");
+
+        public GrpcClientDiagnosticListener(Tracer tracer)
+            : base("Grpc.Net.Client", tracer)
+        {
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnStartActivity";
+            if (!(this.startRequestFetcher.Fetch(payload) is HttpRequestMessage request))
+            {
+                AdapterEventSource.Log.NullPayload(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                return;
+            }
+
+            if (request.Headers.Contains("traceparent"))
+            {
+                // this request is already instrumented, we should back off
+                return;
+            }
+
+            // The path is also available from the grpc.method tag on the activity. Maybe that should be used instead
+            var path = request.RequestUri.AbsolutePath.Trim('/');
+            this.Tracer.StartActiveSpanFromActivity(path, activity, SpanKind.Client, out var span);
+
+            if (span.IsRecording)
+            {
+                var rx = new Regex(@"(?<package>\w+).(?<service>\w+)/(?<method>\w+)", RegexOptions.Compiled);
+                var rpcService = rx.Match(path).Groups["service"].Value;
+
+                span.SetAttribute("rpc.service", rpcService);
+
+                var uriHostNameType = Uri.CheckHostName(request.RequestUri.Host);
+                if (uriHostNameType == UriHostNameType.IPv4 || uriHostNameType == UriHostNameType.IPv6)
+                {
+                    span.SetAttribute("net.peer.ip", request.RequestUri.Host);
+                }
+                else
+                {
+                    span.SetAttribute("net.peer.name", request.RequestUri.Host);
+                }
+
+                span.SetAttribute("net.peer.port", request.RequestUri.Port);
+            }
+
+            // Do we want to propagate trace context here? GrpcClient is built on top of HttpClient.
+            // Currently a span is generated by both GrpcClient and HttpClient. This is probably redundant.
+            // If we do propagation here, then the HttpClient instrumentation would be skipped as it checks if a traceparent header is already present
+        }
+
+        public override void OnStopActivity(Activity activity, object payload)
+        {
+            const string EventNameSuffix = ".OnStopActivity";
+            var span = this.Tracer.CurrentSpan;
+            try
+            {
+                if (span == null || !span.Context.IsValid)
+                {
+                    AdapterEventSource.Log.NullOrBlankSpan(nameof(GrpcClientDiagnosticListener) + EventNameSuffix);
+                    return;
+                }
+
+                if (span.IsRecording)
+                {
+                    // set span.Status
```

Comment:
At this moment we still have the Status on the span so it will be good to try to set it based on some info from the activity - is there any available?

Reply:
Yes, there is a status available. I'll get it in a follow-up commit.